Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -43,250 +43,183 @@ extension BinaryInteger {
+///
+/// - Returns: The `remainder`, which is a value in the range of `0 ..< divisor`.
+///
+private func formQuotientWithRemainderForUnsignedInteger(
+words dividend: Slice<UnsafeMutableBufferPointer<UInt>>, dividingBy divisor: UInt) -> UInt {
+    var remainder = UInt.zero
+    
+    for index in (dividend.startIndex ..< dividend.endIndex).reversed() {
+        (dividend.base[index], remainder) = divisor.dividingFullWidth((high: remainder, low: dividend.base[index]))
     }
+    
+    return remainder as UInt
 }
 
-extension UInt {
-    /// Formats `self` in "Numeric string" format (https://speleotrove.com/decimal/daconvs.html) which is the required input form for certain ICU functions (e.g. `unum_formatDecimal`).
-    ///
-    /// This is intended to be used only by the `numericStringRepresentation` property (both the specialised form below and the generic one for all BinaryIntegers, above).  Prefer the `numericStringRepresentation` property for all other use-cases.
-    ///
-    /// - Parameter intoEndOfBuffer: The buffer to write into, which _must_ contain enough space for the result.  The formatted output is placed into the _end_ of this buffer ("right-aligned", if you will), though the output is numerically still left-to-right.  The contents of this buffer do not have to be pre-initialised.
-    /// - Returns: How many entries (UInt8s) of the buffer were used.  Note that zero is a valid return value, as nothing is written to the buffer if `self` is zero (this may be odd but it's acceptable to `numericStringRepresentation` and it simplifies the overall implementation).
-    fileprivate func numericStringRepresentation(intoEndOfBuffer buffer: inout Slice<UnsafeMutableBufferPointer<UInt8>>) -> Int {
-        guard .zero != self else { return 0 } // Easier to special-case this here than deal with it below (annoying off-by-one potential errors).
-
-        var insertionPoint = buffer.endIndex - 1
-        var tmp = self
-
-        // Keep dividing by ten until the value disappears.  Each time we divide, we get one more digit for the output as the remainder of the division.  Since with this approach digits "pop off" from least significant to most, the output buffer is filled in reverse.
-        while .zero != tmp {
-            let (quotient, remainder) = tmp.quotientAndRemainder(dividingBy: 10)
-
-            buffer[insertionPoint] = UInt8(ascii: "0") + UInt8(remainder)
-
-            if .zero != quotient {
-                insertionPoint -= 1
-                assert(insertionPoint >= buffer.startIndex, "Buffer is too small (\(buffer.count) UInt8s) to contain the result.")
-            }
-
-            tmp = quotient
-        }
-
-        return insertionPoint.distance(to: buffer.endIndex)
+/// Forms the `two's complement` of a binary integer.
+///
+/// - Parameter words: A binary integer's mutable words.
+///
+private func formTwosComplementForBinaryInteger(words: UnsafeMutableBufferPointer<UInt>) {
```

Comment:
I don't have a strong opinion about it, but I found your comment a bit perplexing because I don't ascribe signedness to it. The 2's complement format is named after the 2's complement operation `~x &+ 1` (unsigned, fixed-width, `2 ^ x.bitWidth - x`). It took me a while to pinpoint, but I believe this snippets highlight our different viewpoints:


```swift
for x in Int8.min ... Int8.max {
    var magnitude = UInt8(bitPattern: x) // some bit pattern
    
    if type(of: x).isSigned && magnitude >> (x.bitWidth - 1) == 1 {
        magnitude = ~magnitude &+ 1 // form 2's complement when needed
    }
    
    precondition(x.magnitude == magnitude) // bit pattern is magnitude
}

// MARK: Dunno, but how I think you see it:

for x in Int8.min ... Int8.max {
    let magnitude: UInt8
    
    magnitude = UInt8(bitPattern: x < 0 ? ~x &+ 1 : x) // some representation conversion
    
    precondition(x.magnitude == magnitude)
}
```

Reply:
I think the key point is that you can't fit `x.magnitude` into `Self` for any signed `Self`, specifically because of `type(of: x).min`.  e.g. for `Int8` (just to keep the numbers small) `Int8.min` is -128.  The inverse of that is 128, which is greater than `Int8.max` (of 127).  You have to upgrade to `Int16` to represent it _in two's complement form_.

Aside: this is presumably why the type of the `magnitude` property is _not_ `Self`, but rather an associated value defined by each specific type.  All the built-in signed integer types use their unsigned partners for it.

So, the function in question _cannot_ perform two's complement correctly for all values; it's not a generic two's complement implementation.  What it _can_ do is:

1. Invert any _positive_ two's complement-encoded value.
2. Invert any negative, two's complement-encoded value into its positive form _as an unsigned-type of the same bit width_.

(1) isn't relevant to the function's actual use in this file, only (2).  Thus why I suggested a name that is more specific to its intended use (even if it technically ignores another use it can work for, (1)). 

This is all just to try to clarify.  I don't feel that this is a particular important point in the grand scheme of things, either.  It just seems like a sensible little refinement, to me.

Reply:
Or at least, not for _all_ signed `BinaryInteger`s.  You can certainly have some where that limitation doesn't apply - e.g. because they're flexible-width - or I suppose hypothetically even a `FixedWidthInteger` that constraints it minimum value to something less (in magnitude) than 2^(bitWidth-1).  But the code in question has to work for _all_ `BinaryInteger`s.

Reply:
It performs the *two's complement operation* `~x &+ 1`, which is defined for all binary integers. It's signedness agnostic, and it does not reinterpret anything. Whether the result of `~x &+ 1` is useful to us is determined at the call site.

Reply:
This is only to clarify that you cannot find an input where `~x &+ 1` does not fit. In the one place it's used, the `words` has—to your point—been reinterpreted as an unsigned `magnitude`. The type system isn't there to help, so there's no bit cast reinterpreting it. I get the feeling that you'd be happier with renaming the `magnitude` as `reinterpretedUnsignedMagnitude` or somesuch, instead of renaming this method :man_shrugging:

Reply:
I changed the call site to the following for clarity:

```swift
//  We reinterpret the words as an unsigned binary integer.
var magnitude = /* consume */ words
//  Note that negative values are in two's complement form.
let isLessThanZero = isSigned && Int(bitPattern: magnitude.last ?? .zero) < .zero
//  The **unsigned** magnitude is formed when the words represent a negative value.
if  isLessThanZero {
    formTwosComplementForBinaryInteger002(words: magnitude)
}
```